File: sdks/go/bcs/bcs.go
// Package bcs provides Binary Canonical Serialization for Go.
//
// BCS is a deterministic binary serialization format that guarantees
// canonical representation - every value has exactly one valid encoding.
//
// # Quick Start
//
//	// Serialization
//	ser := bcs.NewSerializer()
//	ser.WriteU8(1)
//	ser.WriteU64(100)
//	ser.WriteString("hello")
//	bytes := ser.Bytes()
//
//	// Deserialization
//	des := bcs.NewDeserializer(bytes)
//	v1, _ := des.ReadU8()
//	v2, _ := des.ReadU64()
//	v3, _ := des.ReadString()
//	des.CheckEnd()
//
// # Supported Types
//
// - Booleans: bool
// - Unsigned integers: u8, u16, u32, u64, u128, u256
// - Signed integers: i8, i16, i32, i64, i128, i256
// - Byte arrays: bytes (length-prefixed)
// - Strings: string (UTF-8, length-prefixed)
// - Fixed bytes: fixed_bytes (no length prefix)
// - Options: option (0x00 for None, 0x01 for Some)
// - Vectors: vector (length-prefixed)
// - Maps: map (length-prefixed, sorted by key bytes)
// - Enums: via variant index
//
// # Constants
//
// - MaxSequenceLength: 2^31 - 1 (2,147,483,647)
// - MaxContainerDepth: 500
package bcs

// SerializeU8 serializes a u8 value.
func SerializeU8(value uint8) []byte {
	return []byte{value}
}

// SerializeU64 serializes a u64 value.
func SerializeU64(value uint64) []byte {
	ser := NewSerializer()
	ser.WriteU64(value)
	return ser.Bytes()
}

// SerializeString serializes a string value.
func SerializeString(value string) []byte {
	ser := NewSerializer()
	ser.WriteString(value)
	return ser.Bytes()
}

// SerializeBytes serializes a bytes value.
func SerializeBytes(value []byte) []byte {
	ser := NewSerializer()
	ser.WriteBytes(value)
	return ser.Bytes()
}

// DeserializeU8 deserializes a u8 value.
func DeserializeU8(data []byte) (uint8, error) {
	des := NewDeserializer(data)
	value, err := des.ReadU8()
	if err != nil {
		return 0, err
	}
	if err := des.CheckEnd(); err != nil {
		return 0, err
	}
	return value, nil
}

// DeserializeU64 deserializes a u64 value.
func DeserializeU64(data []byte) (uint64, error) {
	des := NewDeserializer(data)
	value, err := des.ReadU64()
	if err != nil {
		return 0, err
	}
	if err := des.CheckEnd(); err != nil {
		return 0, err
	}
	return value, nil
}

// DeserializeString deserializes a string value.
func DeserializeString(data []byte) (string, error) {
	des := NewDeserializer(data)
	value, err := des.ReadString()
	if err != nil {
		return "", err
	}
	if err := des.CheckEnd(); err != nil {
		return "", err
	}
	return value, nil
}

// DeserializeBytes deserializes a bytes value.
func DeserializeBytes(data []byte) ([]byte, error) {
	des := NewDeserializer(data)
	value, err := des.ReadBytes()
	if err != nil {
		return nil, err
	}
	if err := des.CheckEnd(); err != nil {
		return nil, err
	}
	return value, nil
}

File: sdks/swift/Sources/BCS/BCS.swift
// Copyright (c) BCS SDK Contributors
// SPDX-License-Identifier: Apache-2.0

import Foundation

/// Binary Canonical Serialization (BCS) - Swift Implementation
///
/// BCS is a deterministic binary serialization format designed for
/// canonical representation of data structures.
///
/// ## Quick Start
///
/// ```swift
/// // Serialize
/// let ser = BcsSerializer()
/// ser.writeU64(12345)
///    .writeString("hello")
///    .writeBool(true)
/// let bytes = ser.toBytes()
///
/// // Deserialize
/// let des = BcsDeserializer(bytes)
/// let num = try des.readU64()
/// let str = try des.readString()
/// let flag = try des.readBool()
/// try des.checkEnd()
/// ```
public enum BCS {
    // MARK: - Convenience Serialization Functions

    /// Serialize a UInt8 value
    public static func serializeU8(_ value: UInt8) -> [UInt8] {
        BcsSerializer().writeU8(value).toBytes()
    }

    /// Serialize a UInt16 value
    public static func serializeU16(_ value: UInt16) -> [UInt8] {
        BcsSerializer().writeU16(value).toBytes()
    }

    /// Serialize a UInt32 value
    public static func serializeU32(_ value: UInt32) -> [UInt8] {
        BcsSerializer().writeU32(value).toBytes()
    }

    /// Serialize a UInt64 value
    public static func serializeU64(_ value: UInt64) -> [UInt8] {
        BcsSerializer().writeU64(value).toBytes()
    }

    /// Serialize a Bool value
    public static func serializeBool(_ value: Bool) -> [UInt8] {
        BcsSerializer().writeBool(value).toBytes()
    }

    /// Serialize a String value
    public static func serializeString(_ value: String) throws -> [UInt8] {
        try BcsSerializer().writeString(value).toBytes()
    }

    /// Serialize bytes
    public static func serializeBytes(_ value: [UInt8]) throws -> [UInt8] {
        try BcsSerializer().writeBytes(value).toBytes()
    }

    // MARK: - Convenience Deserialization Functions

    /// Deserialize a UInt8 value
    public static func deserializeU8(_ data: [UInt8]) throws -> UInt8 {
        let des = BcsDeserializer(data)
        let value = try des.readU8()
        try des.checkEnd()
        return value
    }

    /// Deserialize a UInt16 value
    public static func deserializeU16(_ data: [UInt8]) throws -> UInt16 {
        let des = BcsDeserializer(data)
        let value = try des.readU16()
        try des.checkEnd()
        return value
    }

    /// Deserialize a UInt32 value
    public static func deserializeU32(_ data: [UInt8]) throws -> UInt32 {
        let des = BcsDeserializer(data)
        let value = try des.readU32()
        try des.checkEnd()
        return value
    }

    /// Deserialize a UInt64 value
    public static func deserializeU64(_ data: [UInt8]) throws -> UInt64 {
        let des = BcsDeserializer(data)
        let value = try des.readU64()
        try des.checkEnd()
        return value
    }

    /// Deserialize a Bool value
    public static func deserializeBool(_ data: [UInt8]) throws -> Bool {
        let des = BcsDeserializer(data)
        let value = try des.readBool()
        try des.checkEnd()
        return value
    }

    /// Deserialize a String value
    public static func deserializeString(_ data: [UInt8]) throws -> String {
        let des = BcsDeserializer(data)
        let value = try des.readString()
        try des.checkEnd()
        return value
    }

    /// Deserialize bytes
    public static func deserializeBytes(_ data: [UInt8]) throws -> [UInt8] {
        let des = BcsDeserializer(data)
        let value = try des.readBytes()
        try des.checkEnd()
        return value
    }

    // MARK: - Hex Utilities

    /// Convert bytes to hex string
    public static func bytesToHex(_ bytes: [UInt8]) -> String {
        bytes.map { String(format: "%02x", $0) }.joined()
    }

    /// Convert hex string to bytes
    public static func hexToBytes(_ hex: String) -> [UInt8]? {
        var result: [UInt8] = []
        var index = hex.startIndex

        while index < hex.endIndex {
            guard let nextIndex = hex.index(index, offsetBy: 2, limitedBy: hex.endIndex) else {
                return nil
            }
            guard let byte = UInt8(hex[index..<nextIndex], radix: 16) else {
                return nil
            }
            result.append(byte)
            index = nextIndex
        }

        return result
    }
}
